nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java

```diff
@@ -90,4 +112,178 @@ private static void invalidInstantiationError(
+      if (paramTypedTree.getTypeArguments().isEmpty()) {
+        // no explicit type parameters
+        return;
+      }
+      rhsType =
+          typeWithPreservedAnnotations(
+              (ParameterizedTypeTree) ((NewClassTree) rhsTree).getIdentifier());
+    }
+    if (lhsType != null && rhsType != null) {
+      compareNullabilityAnnotations((Type.ClassType) lhsType, (Type.ClassType) rhsType, tree);
+    }
+  }
+
+  /**
+   * Compare two types from an assignment for identical type parameter nullability, recursively
+   * checking nested generic types. See <a
+   * href="https://jspecify.dev/docs/spec/#nullness-delegating-subtyping">the JSpecify
+   * specification</a> and <a
+   * href="https://docs.oracle.com/javase/specs/jls/se14/html/jls-4.html#jls-4.10.2">the JLS
+   * subtyping rules for class and interface types</a>.
+   *
+   * @param lhsType type for the lhs of the assignment
+   * @param rhsType type for the rhs of the assignment
+   * @param tree tree representing the assignment
+   */
+  private void compareNullabilityAnnotations(
+      Type.ClassType lhsType, Type.ClassType rhsType, Tree tree) {
+    Types types = state.getTypes();
+    // The base type of rhsType may be a subtype of lhsType's base type.  In such cases, we must
+    // compare lhsType against the supertype of rhsType with a matching base type.
+    rhsType = (Type.ClassType) types.asSuper(rhsType, lhsType.tsym);
+    if (rhsType == null) {
+      throw new RuntimeException("did not find supertype of " + rhsType + " matching " + lhsType);
+    }
+    List<Type> lhsTypeArguments = lhsType.getTypeArguments();
+    List<Type> rhsTypeArguments = rhsType.getTypeArguments();
+    if (lhsTypeArguments.size() != rhsTypeArguments.size()) {
+      throw new RuntimeException(
+          "number of types arguments in " + rhsType + " does not match " + lhsType);
+    }
+    for (int i = 0; i < lhsTypeArguments.size(); i++) {
+      Type lhsTypeArgument = lhsTypeArguments.get(i);
+      Type rhsTypeArgument = rhsTypeArguments.get(i);
+      com.sun.tools.javac.util.List<Attribute.TypeCompound> annotationMirrorsLHS =
+          lhsTypeArgument.getAnnotationMirrors();
+      com.sun.tools.javac.util.List<Attribute.TypeCompound> annotationMirrorsRHS =
+          rhsTypeArgument.getAnnotationMirrors();
+      boolean isLHSNullableAnnotated =
+          Nullness.hasNullableAnnotation(annotationMirrorsLHS.stream(), config);
+      boolean isRHSNullableAnnotated =
+          Nullness.hasNullableAnnotation(annotationMirrorsRHS.stream(), config);
+      if (isLHSNullableAnnotated != isRHSNullableAnnotated) {
+        reportInvalidAssignmentInstantiationError(tree, lhsType, rhsType, state, analysis);
+        return;
+      }
+      // nested generics
+      if (lhsTypeArgument.getTypeArguments().length() > 0) {
+        compareNullabilityAnnotations(
+            (Type.ClassType) lhsTypeArgument, (Type.ClassType) rhsTypeArgument, tree);
+      }
+    }
+  }
+
+  /**
```

Comment:
I think we are missing a one-sentence Javadoc before the param and return description. Possibly duplicating some of the explanation from the call-site, but it's good to have that on hand when looking at this method, since it's purpose is not obvious from the name.